### PR TITLE
Integrate Invoke-BloodHound

### LINKED ISSUE #11258: Integrate Invoke-BloodHound
It would be nice to have `Invoke-BloodHound` as a post module.

### PR DESCRIPTION
**Pull request by phra**
This PR will add `Invoke-BloodHound` as post module.

It will fix #11258.

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] Get a meterpreter session on Windows
- [x] `use post/windows/gather/bloodhound`
- [x] `run`
- [x] Get `Invoke-BloodHound` usage

image

### CODE REVIEW

**Comment by OJ** [comment]
So this is using the Powershell extension behind the scenes with the `execute_string` option. This isn't "streamed" so my feeling would be that this command would timeout constantly.

Have you tested this against a domain with a lot of users/computers/etc ?

**Comment by phra** [comment]
@OJ no, i just wrote the base of the module, i thought about the execution timeout too. it should be run asynchronously for sure.

**Comment by sempervictus** [comment]
The original PSH runner, pre meterp extension should be able to do this just fine. I'll pull it in and test, i think i have a short-circuit which prevents loading the extension unless i explicitly set a flag because it never worked quite right for complex PSH scripts for me...
Thanks for getting this wired up @phra

**Comment by phra** [comment]
@sempervictus you can reuse this commit in order to try the `psh_exec`: https://github.com/rapid7/metasploit-framework/pull/11256/commits/63e9f2d9d8557b80b0f2669c829b151d7dbd77f5

i had to switch to the extension because it spawns a lot of channels in order to stage big payloads and often metasploit was crashing due to this. the meterpreter extension is much more stable.

File: modules/post/windows/gather/bloodhound.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      'Arch' => [ 'x86', 'x64' ],
+      'SessionTypes' => [ 'meterpreter' ]
+    )
+  )
+
+    register_options(
+      [
+        OptString.new('CMD', [ true, 'The command to execute (eg. Invoke-BloodHound)', 'Get-Help Invoke-BloodHound' ])
+      ],
+      self.class
```
Comment (bcoles) [inline modules/post/windows/gather/bloodhound.rb]: `self.class` is redundant.

```suggestion
```

File: modules/post/windows/gather/bloodhound.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    super(update_info(
+      info,
+      'Name' => 'BloodHound',
+      'Description' => %q(This module load in memory Invoke-BloodHound.ps1 and executes it.),
+      'License' => MSF_LICENSE,
+      'Author' => [
+        'BloodHound Gang', # Invoke-BloodHound
+        'phra' # MSF Module
+      ],
+      'References' => [''],
```
Comment (bcoles) [inline modules/post/windows/gather/bloodhound.rb]: Link to `https://github.com/BloodHoundAD/BloodHound` perhaps?

Comment (phra) [reply modules/post/windows/gather/bloodhound.rb]: i will add it.

**Comment by phra** [comment]
@bcoles any suggestion about how to integrate this module with the `loot` database?

**Comment by phra** [comment]
my idea was to parse every single ticket and saved separately like they were normal credentials. (it would be nice to have it also for kiwi output like cobalt strike)

**Comment by h00die** [comment]
Not sure how cobal strike does it (personally never used it), but you can store things to the credential database!
Unfortunately it doesn't look like there is a wiki page giving good dev notes on it but if you search around for `create_credential`, that is the function call.
One example (which creates the service so its stored as SMB/445tcp) can be seen here: https://github.com/rapid7/metasploit-framework/blob/76954957c740525cff2db5a60bcf936b4ee06c42/modules/post/windows/gather/hashdump.rb#L67-L105
A simply example I've been working with which doesn't create the service can be found here: https://github.com/rapid7/metasploit-framework/pull/11264#issuecomment-455762574

As for the type (https://github.com/rapid7/metasploit-framework/blob/bf949b7fd2a705a74ca02f66792f80ab41b4c81a/documentation/api/v1/credential_api_doc.rb#L42)

**Comment by acammack-r7** [comment]
This is a duplicate of https://github.com/rapid7/metasploit-framework/pull/10579.

**Pull request closed without merging**